Unable to build xbps with OpenSSL 1.1

I'm attempting to build on Arch Linux and get several errors. I ran `make -k` in an attempt to log as many errors as possible. gcc is version 7.2.0. openssl is 1.1.0.g.

    fetch/common.c: In function ‘fetch_ssl_verify_altname’:
    fetch/common.c:898:3: error: ‘ASN1_STRING_data’ is deprecated [-Werror=deprecated-declarations]
       ns = (const char *)ASN1_STRING_data(name->d.ia5);
       ^~
    In file included from /usr/include/openssl/bn.h:31:0,
                     from /usr/include/openssl/asn1.h:24,
                     from /usr/include/openssl/objects.h:916,
                     from /usr/include/openssl/evp.h:27,
                     from /usr/include/openssl/x509.h:23,
                     from /usr/include/openssl/x509v3.h:14,
                     from fetch/common.c:62:
    /usr/include/openssl/asn1.h:553:1: note: declared here
     DEPRECATEDIN_1_1_0(unsigned char *ASN1_STRING_data(ASN1_STRING *x))
     ^
    cc1: all warnings being treated as errors
    make[1]: *** [Makefile:62: fetch/common.o] Error 1
    pubkey2fp.c: In function ‘xbps_pubkey2fp’:
    pubkey2fp.c:68:13: error: storage size of ‘mdctx’ isn’t known
      EVP_MD_CTX mdctx;
                 ^~~~~
    pubkey2fp.c:94:27: error: dereferencing pointer to incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
      if (EVP_PKEY_type(pPubKey->type) != EVP_PKEY_RSA) {
                               ^~
    In file included from /usr/include/openssl/asn1.h:24:0,
                     from /usr/include/openssl/objects.h:916,
                     from /usr/include/openssl/evp.h:27,
                     from pubkey2fp.c:16:
    pubkey2fp.c:107:26: error: dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}’
      nLen = BN_num_bytes(pRsa->n);
                              ^
    pubkey2fp.c:143:2: error: implicit declaration of function ‘EVP_MD_CTX_cleanup’; did you mean ‘EVP_MD_CTX_create’? [-Werror=implicit-function-declaration]
      EVP_MD_CTX_cleanup(&mdctx);
      ^~~~~~~~~~~~~~~~~~
      EVP_MD_CTX_create
    pubkey2fp.c:143:2: error: nested extern declaration of ‘EVP_MD_CTX_cleanup’ [-Werror=nested-externs]
    pubkey2fp.c:68:13: error: unused variable ‘mdctx’ [-Werror=unused-variable]
      EVP_MD_CTX mdctx;
                 ^~~~~
    cc1: all warnings being treated as errors
    make[1]: *** [Makefile:72: pubkey2fp.o] Error 1
    make[1]: Target 'all' not remade because of errors.
    make: *** [Makefile:15: all] Error 1

It does, but there are still two errors ("dereferencing pointer to incomplete type"):

    pubkey2fp.c: In function ‘xbps_pubkey2fp’:
    pubkey2fp.c:68:13: error: storage size of ‘mdctx’ isn’t known
      EVP_MD_CTX mdctx;
                 ^~~~~
    pubkey2fp.c:94:27: error: dereferencing pointer to incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
      if (EVP_PKEY_type(pPubKey->type) != EVP_PKEY_RSA) {
                               ^~
    In file included from /usr/include/openssl/asn1.h:24:0,
                     from /usr/include/openssl/objects.h:916,
                     from /usr/include/openssl/evp.h:27,
                     from pubkey2fp.c:16:
    pubkey2fp.c:107:26: error: dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}’
      nLen = BN_num_bytes(pRsa->n);
                              ^
    pubkey2fp.c:143:2: warning: implicit declaration of function ‘EVP_MD_CTX_cleanup’; did you mean ‘EVP_MD_CTX_create’? [-Wimplicit-function-declaration]
      EVP_MD_CTX_cleanup(&mdctx);
      ^~~~~~~~~~~~~~~~~~
      EVP_MD_CTX_create
    pubkey2fp.c:143:2: warning: nested extern declaration of ‘EVP_MD_CTX_cleanup’ [-Wnested-externs]
    pubkey2fp.c:68:13: warning: unused variable ‘mdctx’ [-Wunused-variable]
      EVP_MD_CTX mdctx;
                 ^~~~~
    make[1]: *** [Makefile:72: pubkey2fp.o] Error 1
    make[1]: Target 'all' not remade because of errors.
    make: *** [Makefile:15: all] Error 1

On Sat, Nov 25, 2017 at 10:22:40AM +0000, James “Wes” Barnett wrote:
>     pubkey2fp.c:143:2: warning: implicit declaration of function ‘EVP_MD_CTX_cleanup’; did you mean ‘EVP_MD_CTX_create’? [-Wimplicit-function-declaration]
>       EVP_MD_CTX_cleanup(&mdctx);

No longer available in openssl 1.1+.

--- a/README.md
+++ b/README.md
@@ -39,7 +39,7 @@ To build this you'll need:
   - [GNU make](http://www.gnu.org/software/make/)
   - [pkg-config](http://www.freedesktop.org/wiki/Software/pkg-config/)
   - [zlib](http://www.zlib.net)
-  - [openssl](http://www.openssl.org)
+  - [openssl < 1.1](http://www.openssl.org) or [libressl](https://www.libressl.org/)
   - [libarchive >= 3.2.0](http://www.libarchive.org)
 
 and optionally:

--- a/configure
+++ b/configure
@@ -675,43 +675,21 @@ else
 fi
 
 #
-# OpenSSL libssl with pkg-config support is required when building
-# the static binaries.
+# libssl with pkg-config support is required.
 #
-if [ "$BUILD_STATIC" = "yes" ]; then
-	printf "Checking for OpenSSL via pkg-config ... "
-	if ! $PKGCONFIG_BIN --exists libssl; then
-		echo "libssl.pc file not found, exiting."
-		exit 1
-	else
-		echo "found version $($PKGCONFIG_BIN --modversion libssl)."
-		echo "CFLAGS += $($PKGCONFIG_BIN --cflags libssl)" >>$CONFIG_MK
-		echo "LDFLAGS +=        $($PKGCONFIG_BIN --libs libssl)" >>$CONFIG_MK
-		echo "STATIC_LIBS +=    $($PKGCONFIG_BIN --libs --static libssl)" \
-			>>$CONFIG_MK
-	fi
+printf "Checking for libssl via pkg-config ... "
+if $PKGCONFIG_BIN --exists 'libssl < 1.1' && ! $PKGCONFIG_BIN --exists libtls ; then
+	echo "found OpenSSL version $($PKGCONFIG_BIN --modversion libssl)."
+elif $PKGCONFIG_BIN --exists libssl libtls; then
+	echo "found LibreSSL version $($PKGCONFIG_BIN --modversion libssl)."
 else
-	func=SHA256
-	printf "Checking for OpenSSL $func() ... "
-cat <<EOF > _$func.c
-#include <openssl/sha.h>
-int main(void) {
-	SHA256(NULL, NULL, 0);
-	return 0;
-}
-EOF
-	if $XCC _$func.c -lcrypto -o _$func 2>/dev/null; then
-		echo yes.
-		echo "LDFLAGS +=	-lcrypto" >>$CONFIG_MK
-	else
-		CRYPTO_SHA256=no
-	fi
-	rm -f _$func.c _$func
-	if [ "$CRYPTO_SHA256" = "no" ]; then
-		echo "Failed to link with your system's OpenSSL library, can't continue..."
-		exit 1
-	fi
+	echo "libssl.pc file not found or incompatible version detected, exiting."
+	exit 1
 fi
+echo "CFLAGS += $($PKGCONFIG_BIN --cflags libssl)" >>$CONFIG_MK
+echo "LDFLAGS +=        $($PKGCONFIG_BIN --libs libssl)" >>$CONFIG_MK
+echo "STATIC_LIBS +=    $($PKGCONFIG_BIN --libs --static libssl)" \
+	>>$CONFIG_MK
 
 #
 # If --enable-static enabled, build static binaries.

